### PR TITLE
Enable support to get openflow count from all HVs

### PR DESCRIPTION
**Pull request by noah8713**

### CODE REVIEW

File: rally_ovs/plugins/ovs/scenarios/ovn.py [hunk 1]
```diff
@@ -413,7 +413,17 @@ def _wait_up_port(self, lports, wait_sync, install_method):
...
+        sandbox_num = len(sandboxes)
+        install_method = self.install_method
+        for j in range(sandbox_num):
+            sandbox = sandboxes[j]["name"]
+            farm = sandboxes[j]["farm"]
+            ovs_ofctl = self.farm_clients(farm, "ovs-ofctl")
+            ovs_ofctl.set_sandbox(sandbox, install_method)
+            bridge = sandbox_args.get('bridge', 'br-int')
+            lflow_count = ovs_ofctl.dump_flows(bridge)
+            LOG.debug('openflow count on %s is %s' % (sandbox, lflow_count))
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/scenarios/ovn.py]: How does it work without return value? Just debug log?

File: rally_ovs/plugins/ovs/ovsclients_impl.py [hunk 2]
```diff
@@ -289,6 +289,82 @@ def db_set(self, table, record, *col_values):
             self.run("set", args=args)
 
     def create_client(self):
-        print "*********   call OvnNbctl.create_client"
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/ovsclients_impl.py]: Unrelated change. Please submit a new patch.

File: rally_ovs/plugins/ovs/ovsclients_impl.py [hunk 3]
```diff
@@ -289,6 +289,82 @@ def db_set(self, table, record, *col_values):
...
+    class _OvsOfctl(object):
+
+        def __init__(self, credential):
+            self.ssh = get_ssh_from_credential(credential)
+            self.context = {}
+            self.batch_mode = False
+            self.sandbox = None
+            self.cmds = None
+
+        def enable_batch_mode(self, value=True):
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/ovsclients_impl.py]: ofctl doesn't have batch mode. please remove all batch mode related code.

File: rally_ovs/plugins/ovs/ovsclients_impl.py [hunk 4]
```diff
@@ -289,6 +289,82 @@ def db_set(self, table, record, *col_values):
...
+            self.batch_mode = bool(value)
+
+        def set_sandbox(self, sandbox, install_method="sandbox"):
+            self.sandbox = sandbox
+            self.install_method = install_method
+
+        def run(self, cmd, opts=[], args=[], stdout=sys.stdout, stderr=sys.stderr):
+            self.cmds = self.cmds or []
+
+            # TODO: tested with non batch_mode only for docker
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/ovsclients_impl.py]: This is misleading. Please don't add docker related code if not tested. A TODO comment can be added for supporting docker in the future.

**Review by hzhou8 — CHANGES_REQUESTED** [review]

File: rally_ovs/plugins/ovs/ovsclients_impl.py (hunk 4 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/ovsclients_impl.py]: Ok will do

File: rally_ovs/plugins/ovs/ovsclients_impl.py (hunk 3 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/ovsclients_impl.py]: Sure!

File: rally_ovs/plugins/ovs/ovsclients_impl.py (hunk 2 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/ovsclients_impl.py]: ok I will address that in separate patch

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 1 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: Yeah just debug logs for now. Can extend it to write to file?

File: rally_ovs/plugins/ovs/scenarios/ovn.py [hunk 5]
```diff
@@ -413,7 +413,17 @@ def _wait_up_port(self, lports, wait_sync, install_method):
         if wait_sync != "none":
             ovn_nbctl.sync(wait_sync)
 
-
-
-
+    @atomic.action_timer("ovn_network.list_oflow_count_on_all_sandboxes")
+    def _list_oflow_count_on_all_sandboxes(self, sandboxes,
+                                           sandbox_args):
```
Comment (jsitnicki) [inline rally_ovs/plugins/ovs/scenarios/ovn.py]: Name is a bit confusing. We're actually listing flow count on given sandboxes, right? Not all sandboxes in the deployment.

File: rally_ovs/plugins/ovs/scenarios/ovn.py [hunk 6]
```diff
@@ -413,7 +413,17 @@ def _wait_up_port(self, lports, wait_sync, install_method):
...
+        sandbox_num = len(sandboxes)
+        install_method = self.install_method
+        for j in range(sandbox_num):
```
Comment (jsitnicki) [inline rally_ovs/plugins/ovs/scenarios/ovn.py]: Why not just iterate over values and avoid `sandbox_num` all together. I think that would be more idiomatic.

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 1 above)
Comment (jsitnicki) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: Scenario can report arbitrary data, which becomes part of the report:

https://rally.readthedocs.io/en/latest/task/index.html#tab-scenario-data

I would suggest looking into using this API:

https://github.com/openstack/rally/blob/master/rally/task/scenario.py#L146

Logs are not really intended for this.

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 5 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: @jsitnicki it's for each SBs.

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 6 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: Sure that sounds fine too! :)

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 1 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: @jsitnicki  Ack! Will give it a try

Comment (noah8713) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: Done!

File: rally_ovs/plugins/ovs/ovsclients_impl.py (hunk 2 above)
Comment (noah8713) [reply rally_ovs/plugins/ovs/ovsclients_impl.py]: Sure! will submit in separate patch

**Comment by noah8713** [comment]
Tested with creating 2 sandboxes and one controller which created 6 total openflows on each sandbox.

Works as expected below:
--------------------------------------------------------------------------------
Task 6eed038b-ac5f-49e8-913d-d741cd1e5e96: finished
--------------------------------------------------------------------------------

test scenario OvnNetwork.list_oflow_count_on_all_sandboxes
args position 0
args values:
{
  "runner": {
    "type": "serial",
    "times": 1
  },
  "context": {
    "sandbox": {},
    "ovn_multihost": {
      "controller": "ovn-new-controller-node"
    }
  }
}
+---------------------------------------------------------------------------------------------------------------------+
|                                                Response Times (sec)                                                 |
+------------------------------------------------+-------+--------+--------+--------+-------+-------+---------+-------+
| action                                         | min   | median | 90%ile | 95%ile | max   | avg   | success | count |
+------------------------------------------------+-------+--------+--------+--------+-------+-------+---------+-------+
| ovn_network.list_oflow_count_for_all_sandboxes | 0.354 | 0.354  | 0.354  | 0.354  | 0.354 | 0.354 | 100.0%  | 1     |
| total                                          | 0.361 | 0.361  | 0.361  | 0.361  | 0.361 | 0.361 | 100.0%  | 1     |
+------------------------------------------------+-------+--------+--------+--------+-------+-------+---------+-------+

Scenario Specific Results

+-------------------------------------------------------------------+
|           Openflow count on each sandbox in StackedArea           |
+----------------------+-----+--------+--------+--------+-----+-----+
| key                  | min | median | 90%ile | 95%ile | max | avg |
+----------------------+-----+--------+--------+--------+-----+-----+
| sandbox-192.230.64.0 | 6.0 | 6.0    | 6.0    | 6.0    | 6.0 | 6.0 |
| sandbox-192.230.64.1 | 6.0 | 6.0    | 6.0    | 6.0    | 6.0 | 6.0 |
+----------------------+-----+--------+--------+--------+-----+-----+

Load duration: 0.366648197174
Full duration: 0.391592979431

File: rally_ovs/plugins/ovs/scenarios/ovn.py (hunk 5 above)
Comment (jsitnicki) [reply rally_ovs/plugins/ovs/scenarios/ovn.py]: OK, I think I follow now.

The scenario processes all sandboxes (`OvnNetwork.list_oflow_count_on_all_sandboxes`).

The atomic action processes the sandboxes passed in the arguments (`Ovn._list_oflow_count_for_all_sandboxes`)

**Comment by jsitnicki** [comment]
Awesome, LGTM!

**Comment by jsitnicki** [comment]
Actually, one more comment.

I would consider if this new scenario doesn't fit better into `OvnSandbox` module.

**Comment by noah8713** [comment]
Thanks @jsitnicki for review.
You mean move the lflow logic to ovn_sandbox? The reason I choose current place is because getting openflow count is related to ovs scenarios, similar to other ovn/ovs client commands. Hence, I feel this would be ok. Let me know

File: rally_ovs/plugins/ovs/ovsclients_impl.py [hunk 7]
```diff
@@ -292,3 +292,67 @@ def create_client(self):
...
+            self.cmds = None
+
+
+        def set_sandbox(self, sandbox, install_method="sandbox"):
+            self.sandbox = sandbox
+            self.install_method = install_method
+
+        def run(self, cmd, opts=[], args=[], stdout=sys.stdout, stderr=sys.stderr):
+            # TODO: add support for docker
+            self.cmds = self.cmds or []
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/ovsclients_impl.py]: no batch mode for ovs-ofctl, so self.cmds is useless for this class. The run() function should simply self.ssh.run().

File: rally_ovs/plugins/ovs/ovsclients_impl.py [hunk 8]
```diff
@@ -292,3 +292,67 @@ def create_client(self):
...
+
+            cmd = itertools.chain(["ovs-ofctl"], opts, [cmd], args)
+            self.cmds.append(" ".join(cmd))
+
+            self.ssh.run("\n".join(self.cmds),
+                         stdout=stdout, stderr=stderr)
+
+            self.cmds = None
+
+        def flush(self):
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/ovsclients_impl.py]: this function is not needed because there is no batch mode

File: rally_ovs/plugins/ovs/scenarios/ovn_network.py [hunk 9]
```diff
@@ -114,4 +114,10 @@ def bind_ports(self):
     def bind_and_unbind_ports(self):
         pass
 
-
+    @scenario.configure(context={})
+    def list_oflow_count_for_all_sandboxes(self, sandbox_args={}):
```
Comment (hzhou8) [inline rally_ovs/plugins/ovs/scenarios/ovn_network.py]: Better to rename the function as list_oflow_count_for_sandboxes, since it doesn't have to be all. It depends on the sandbox_args. Same for L417.

**Review by hzhou8 — CHANGES_REQUESTED** [review]

**Review by hzhou8 — APPROVED** [review]
